File: src/backend/identityaccess/src/main/java/org/taonaw/identityaccess/domain/model/shared/PlainTextPassword.java
package org.taonaw.identityaccess.domain.model.shared;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import org.taonaw.identityaccess.domain.shared.Assertion;

@Getter(value = AccessLevel.PACKAGE)
@EqualsAndHashCode
public class PlainTextPassword {
    private final String value;

    public PlainTextPassword(@NonNull String value) {
        Assertion.argumentRange(value, 8, 20);
        Assertion.argumentPatternMatches(value, "[a-zA-Z0-9]*", "半角英数でなければなりません。");
        this.value = value;
    }

    public Password encode(@NonNull IPasswordEncoder passwordEncoder) {
        return new Password(passwordEncoder.encode(value));
    }
}

File: src/backend/identityaccess/src/main/kotlin/org/taonaw/identityaccess/domain/model/shared/PlainTextPassword.kt
package org.taonaw.identityaccess.domain.model.shared

data class PlainTextPassword(private val value: String) {

    companion object {
        const val LENGTH_MIN = 8
        const val LENGTH_MAX = 20
        const val FORMAT_PATTERN = "[a-zA-Z0-9]*"
    }

    init {
        require(value.length in LENGTH_MIN..LENGTH_MAX) {
            "パスワードは $LENGTH_MIN 文字以上 $LENGTH_MAX 文字以下でなければなりません。"
        }
        require(Regex(FORMAT_PATTERN).matches(value)) {
            "パスワードは半角英数でなければなりません。"
        }
    }

    fun hash(passwordHashingService: PasswordHashingService): HashedPassword {
        return passwordHashingService.hash(value)
    }

    fun matches(hashedPassword: HashedPassword, passwordHashingService: PasswordHashingService): Boolean {
        return passwordHashingService.matches(value, hashedPassword)
    }
}